macros/datediff.sql

```diff
@@ -0,0 +1,43 @@
+{%- macro datediff(part, left_val, right_val, date_format='%Y-%m-%d') -%}
+    {# determines the delta (in `part` units) between first_val and second_val.
+       *Note* the order of left_val, right_val is reversed from Snowflake.
+       ARGS:
+         - part (string) one of 'day', 'week', 'month', 'year'
+         - left_val (date/timestamp) the value before the minus in the equation "left - right"
+         - right_val (date/timestamp) the value after the minus in the equation "left - right"
+         - date_format (pattern) a string pattern for the provided arguments (primarily for BigQuery)
+       RETURNS: An integer representing the delta in `part` units
+    #}
+
+    {%if part not in ('day','week','month','year',) %}
+        {{exceptions.raise_compiler_error("macro datediff for target does not support date part value " ~ part)}}
+    {%- endif -%}
+
```

Comment:
Worth adding a test to see if coercing to date fails and throw an error?

Reply:
Yep and actually bad values too. Good call

Reply:
I don't know that DBT has a clean way to test for exceptions? if they do it's new. gotta dig a little, I'll add that in a later PR

Reply:
Update - there's no effective method to test _postitive_ for exceptions at runtime in DBT :( that's a shame, but I'm not seeing an easy way to test these :( :( . We could build a runtime pytest suite (wouldn't take much) but that's some scope ➕ ➕  for sure.